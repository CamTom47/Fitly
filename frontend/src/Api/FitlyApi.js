import axios from 'axios';

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

/** Fitly API Class
 * 
 * Static class typing together methods used to get/send to the Fitly API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't be any API-aware stuff elsehwere in the frontend
 */

class FitlyApi {
    //the token for interaction with API will be stored here.
    static token;

    static async request( endpoint, data = {}, method = "get"){
        console.debug("API Call:", endpoint, data, method)

        const url = `${BASE_URL}/${endpoint}`;
        let headers;
        
        (FitlyApi.token !== "") ? headers = { Authorization: `Bearer ${FitlyApi.token}` } : headers = null;
        const params = (method === "get")
            ? data
            : {};

            try{
                return (await axios({url, method, data, params, headers})).data;
            } catch(err) {
                console.error("API Error:", err.response);
                console.log(headers)
                let message = err.response.data.error.message;
                throw Array.isArray(message) ? message : [message];
            }
            }


            //Individual Fitly API routes

            //Front End user methods

            /**
             * Get a user's information
             * 
             * Parameter: {data}
             * 
             * {data: {token}}
             * 
             * Returns: 
             *  {"user": {
             *      "username",
             *      "firstName",
             *      "lastName",
             *      "password"
             *      "email",
        *           "isAdmin"}}
             */

            static async findUser(username){
                let res = await this.request(`users/${username}`)
                return res.user
            }

        /**
         * Authenticate a user and login
         * 
         * @param {*} data
         * 
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async login(data){   
            let res = await this.request('auth/token', data, 'post')
            return res.token;
        }

        /** 
         * Sign up a user and create an account
         * 
         * @param {*} data
         *
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async signup(data){
            let res = await this.request(`auth/register`, data, 'post')
            return res.token;
        }

        static async updateUser(username, data){
            let res = await this.request(`users/${username}`, data, 'patch');
            return res.user;
        }

        //Workout Methods
        
        /**
         * Find all workouts in the database that a user has created
         * 
         * 
         */

        static async findAllWorkouts(){
            let res = await this.request(`workouts/`);
            return res.workouts;

        }
        
        
        /**
         * 
         * @param {*} data 
         * {data : {user_id}}
         * 
         * @returns {workout}
         */

        static async findWorkout(data){
            console.log(data)
            let res = await this.request(`workouts/${data.workoutId}`);
            return res.workout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async createWorkout(data){
            let res = await this.request(`workouts/`, data, 'post');
            return res.newWorkout;
        }

        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async updateWorkout(workout_id,data){
            let res = await this.request(`workouts/${workout_id}`, data, 'patch');
            return res.updatedWorkout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteWorkout(workoutId){
            let res = await this.request(`workouts/${workoutId}`, {}, 'delete');
            return res.workout;
        }

        static async addWorkoutCircuit(data){
            let res = await this.request(`workouts//${data.workoutId}/circuits/${data.circuitId}`, data, 'post');
            return res.workoutCircuit;
        }

        //Exercise Methods

        /**
         * Find all Exercises created by a user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercises}
         */

        static async findAllExercises(data){
            let res = await this.request(`exercises/`, data, 'get');
            return res.exercises
        }
        

        /**
         * Find a specfic exercise created by a user
         * @param {*} data 
         * {data {user_id}}
         * @returns {exercise}
         */

        static async findExercise(data){
            let res = await this.request(`exercises/${data.exercise_id}`, data, 'get');
            return res.exercise
        }
        
        /**
         * Creates a new exercise
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercise}
         */

        static async createExercise(data){
            let exercise = await this.request(`exercises/`, data, 'post');
            return exercise
        }
        
        /**
         * Update an existing exercise
         * @param {*} data 
         * {data: {exercise}}
         * @returns {exercise}
         */
        static async updateExercise(exerciseId, data){
            let res = await this.request(`exercises/${exerciseId}`, data, 'patch');
            return res.updatedExercise
        }
        
        /**
         * Delete an existing exercise 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteExercise(data){
            let res = await this.request(`exercises/${data.exercise_id}`, data, 'delete');
            return res.message
        }

        //Category Methods

        /**
         * Get all default categories and categories created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {categories}
         */

        static async findAllCategories(data){
            let res = await this.request(`categories/`, data, 'get');
            return res.categories
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */
        static async findCategory(data){
            let res = await this.request(`categories/${data.category_id}`, data, 'get');
            return res.category;
        }

        /**
         * Create a new category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async createCategory(data){
            let res = await this.request(`categories/`, data, 'post');
            return res.category;
        }

        /**
         * Update an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async updateCategory(data){
            let res = await this.request(`categories/:category_id`, data, 'patch');
            return res.category;
        }

        /**
         * Delete an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteCategory(data){
            let res = await this.request(`categories/:category_id`, data, 'delete');
            return res.category;
        }

        //Equipment Methods

        /**
         * Get all default equipments and equipment created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipments}
         */

        static async findAllEquipments(data){
            let res = await this.request(`equipments/`, data, 'get');
            return res.equipments
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */
        static async findEquipment(data){
            let res = await this.request(`equipments/${data.equipment_id}`, data, 'get');
            return res.equipment;
        }

        /**
         * Create a new equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async createEquipment(data){
            let res = await this.request(`equipments/`, data, 'post');
            return res.equipment;
        }

        /**
         * Update an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async updatEquipment(data){
            let res = await this.request(`equipments/:equipment_id`, data, 'patch');
            return res.equipment;
        }

        /**
         * Delete an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteEquipment(data){
            let res = await this.request(`equipments/:equipment_id`, data, 'delete');
            return res.equipment;
        }

        //Circuit Methods


        static async findAllCircuits(data){
            let res = await this.request(`circuits/`, data, 'get');
            return res.circuits;
        }

        static async findCircuit(data){
            let res = await this.request(`circuits/${data.circuit_id}`, data, 'get');
            return res.circuit;
        }

        static async addCircuit(data){
            let res = await this.request(`circuits/`, data, 'post');
            return res.circuit;
        }
        
        static async updateCircuit(circuitId,data){
            let res = await this.request(`circuits/${circuitId}`, data, 'patch');
            return res.updatedCircuit;
        }

        static async deleteCircuit(circuitId){
            let res = await this.request(`circuits/${circuitId}`, {} , 'delete');
            return res.message;
        }

        static async addExerciseCircuit(data){
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'post');
            return res.circuitExercise;
        }
        
        static async updateExerciseCircuit(data){
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'patch');
            return res.circuitExercise;
        }

        static async findAllMuscleGroups(data){
            let res = await this.request(`muscleGroups/`, data, 'get');
            return res.muscleGroups
        }
        
        static async findMuscleGroup(data){
            let res = await this.request(`muscleGroups/${data.muscleGroupId}`, data, 'get');
            return res.muscleGroup
        }
    }

    FitlyApi.token = ""

export default FitlyApi;