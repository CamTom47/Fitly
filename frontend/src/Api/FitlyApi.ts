import axios, {Axios, AxiosHeaders, AxiosRequestConfig, AxiosResponse, AxiosResponseHeaders, RawAxiosRequestHeaders} from 'axios';

const BASE_URL : string = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

/** Fitly API Class
 * 
 * Static class typing together methods used to get/send to the Fitly API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't be any API-aware stuff elsehwere in the frontend
 */

class FitlyApi {
    //the token for interaction with API will be stored here.
    static token: string;

    static async request( endpoint: string, data: object = {}, method: string = "get"){
        console.debug("API Call:", endpoint, data, method)

        const url: string =  `${BASE_URL}/${endpoint}`
        const params: {} = (method === "get")
        ? data
        : {}
        const headers : {} | undefined = (FitlyApi.token !== "") ? { Authorization: `Bearer ${FitlyApi.token}` } : undefined;

            try{
                const response = (await axios({ url, params, method, data, headers})).data

                return response
            } catch(err) {
                console.error("API Error:", err.response);
                let message: string = err.response.data.error.message;
                throw Array.isArray(message) ? message : [message];
            }
            }


            //Individual Fitly API routes

            //Front End user methods

            /**
             * Get a user's information
             * 
             * Parameter: {data}
             * 
             * {data: {token}}
             * 
             * Returns: 
             *  {"user": {
             *      "username",
             *      "firstName",
             *      "lastName",
             *      "password"
             *      "email",
        *           "isAdmin"}}
             */

            static async findUser<Promise>(username: string) : Promise {
                console.log(username)
                let res = await this.request(`users/${username}`)
                return res.user
            }

        /**
         * Authenticate a user and login
         * 
         * @param {*} data
         * 
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async login<Promise>(data: {}) : Promise{   
            console.log(data)
            let res = await this.request('auth/token', data, 'post')
            return res.token;
        }

        /** 
         * Sign up a user and create an account
         * 
         * @param {*} data
         *
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async signup<Promise>(data: {}) : Promise{
            let res = await this.request(`auth/register`, data, 'post')
            return res.token;
        }

        static async updateUser<Promise>(username: string, data : {}) : Promise{
            let res = await this.request(`users/${username}`, data, 'patch');
            return res.user;
        }

        //Workout Methods
        
        /**
         * Find all workouts in the database that a user has created
         * 
         * 
         */

        static async findAllWorkouts<Promise>(data? : {}) : Promise{
            let res = await this.request(`workouts/`, data, 'get');
            return res.workouts;

        }
        
        
        /**
         * 
         * @param {*} data 
         * {data : {user_id}}
         * 
         * @returns {workout}
         */

        static async findWorkout<Promise>(data: {workoutId: number}) : Promise{
            console.log(data)
            let res = await this.request(`workouts/${data.workoutId}`);
            return res.workout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async createWorkout<Promise>(data : {}) : Promise{
            console.log(data)
            let res = await this.request(`workouts/`, data, 'post');
            return res.newWorkout;
        }

        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async updateWorkout<Promise>(workoutId: number ,data : {}) : Promise{
            let res = await this.request(`workouts/${workoutId}`, data, 'patch');
            return res.updatedWorkout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteWorkout<Promise>(workoutId: number) : Promise{
            let res = await this.request(`workouts/${workoutId}`, {}, 'delete');
            return res.workout;
        }

        static async addWorkoutCircuit<Promise>(data : {workoutId: number | undefined, circuitId: number | undefined}) : Promise{
            let res = await this.request(`workouts//${data.workoutId}/circuits/${data.circuitId}`, data, 'post');
            return res.workoutCircuit;
        }

        //Exercise Methods

        /**
         * Find all Exercises created by a user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercises}
         */

        static async findAllExercises<Promise>(data?: {}) : Promise{
            let res = await this.request(`exercises/`, data, 'get');
            return res.exercises
        }
        

        /**
         * Find a specfic exercise created by a user
         * @param {*} data 
         * {data {user_id}}
         * @returns {exercise}
         */

        static async findExercise<Promise>(data : {exerciseId : number}): Promise{
            let res = await this.request(`exercises/${data.exerciseId}`, data, 'get');
            return res.exercise
        }
        
        /**
         * Creates a new exercise
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercise}
         */

        static async createExercise<Promise>(data: {}): Promise{
            let exercise = await this.request(`exercises/`, data, 'post');
            return exercise
        }
        
        /**
         * Update an existing exercise
         * @param {*} data 
         * {data: {exercise}}
         * @returns {exercise}
         */
        static async updateExercise<Promise>(exerciseId : number, data: {}): Promise{
            let res = await this.request(`exercises/${exerciseId}`, data, 'patch');
            return res.updatedExercise
        }
        
        /**
         * Delete an existing exercise 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteExercise<Promise>(data : {exerciseId: number}): Promise{
            let res = await this.request(`exercises/${data.exerciseId}`, data, 'delete');
            return res.message
        }

        //Category Methods

        /**
         * Get all default categories and categories created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {categories}
         */

        static async findAllCategories<Promise>(data? : {}) : Promise{
            let res = await this.request(`categories/`, data, 'get');
            return res.categories
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */
        static async findCategory<Promise>(data : {categoryId : string}) : Promise{
            let res = await this.request(`categories/${data.categoryId}`, data, 'get');
            return res.category;
        }

        /**
         * Create a new category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async createCategory<Promise>(data : {}) : Promise{
            let res = await this.request(`categories/`, data, 'post');
            return res.category;
        }

        /**
         * Update an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async updateCategory<Promise>(data : {categoryId : number,
                                                    name: string,
                                                    description: string
        }) : Promise{
            let res = await this.request(`categories/${data.categoryId}`, data, 'patch');
            return res.category;
        }

        /**
         * Delete an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteCategory<Promise>(data : {categoryId : number}) : Promise{
            let res = await this.request(`categories/${data.categoryId}`, data, 'delete');
            return res.category;
        }

        //Equipment Methods

        /**
         * Get all default equipments and equipment created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipments}
         */

        static async findAllEquipments<Promise>(data? : {}) : Promise{
            let res = await this.request(`equipments/`, data, 'get');
            return res.equipments
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */
        static async findEquipment<Promise>(data : {equipmentId : number}) : Promise{
            let res = await this.request(`equipments/${data.equipmentId}`, data, 'get');
            return res.equipment;
        }

        /**
         * Create a new equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async createEquipment<Promise>(data : {}) : Promise{
            let res = await this.request(`equipments/`, data, 'post');
            return res.equipment;
        }

        /**
         * Update an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async updatEquipment<Promise>(data : {equipmentId}) : Promise{
            let res = await this.request(`equipments/${data.equipmentId}`, data, 'patch');
            return res.equipment;
        }

        /**
         * Delete an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteEquipment<Promise>(data : {equipmentId : number}) : Promise{
            let res = await this.request(`equipments/${data.equipmentId}`, data, 'delete');
            return res.equipment;
        }

        //Circuit Methods


        static async findAllCircuits<Promise>(data? : {}) : Promise{
            let res = await this.request(`circuits/`, data, 'get');
            return res.circuits;
        }

        static async findCircuit<Promise>(data : {circuitId}) : Promise{
            let res = await this.request(`circuits/${data.circuitId}`, data, 'get');
            return res.circuit;
        }

        static async addCircuit<Promise>(data : {}) : Promise{
            let res = await this.request(`circuits/`, data, 'post');
            return res.circuit;
        }
        
        static async updateCircuit<Promise>(circuitId : number, data : {}) : Promise{
            let res = await this.request(`circuits/${circuitId}`, data, 'patch');
            return res.updatedCircuit;
        }

        static async deleteCircuit<Promise>(circuitId : number) : Promise{
            let res = await this.request(`circuits/${circuitId}`, {} , 'delete');
            return res.message;
        }

        static async addExerciseCircuit<Promise>(data : {circuitId : number | undefined, exerciseId : number | undefined}) : Promise{
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'post');
            return res.circuitExercise;
        }
        
        static async updateExerciseCircuit<Promise>(data : {circuitId: number | undefined, exerciseId: number | undefined}) : Promise{
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'patch');
            return res.circuitExercise;
        }

        static async findAllMuscleGroups<Promise>(data? : {}) : Promise{
            let res = await this.request(`muscleGroups/`, data, 'get');
            return res.muscleGroups
        }
        
        static async findMuscleGroup<Promise>(data : {muscleGroupId : number}) : Promise{
            let res = await this.request(`muscleGroups/${data.muscleGroupId}`, data, 'get');
            return res.muscleGroup
        }
    }

    FitlyApi.token = ""

export default FitlyApi;