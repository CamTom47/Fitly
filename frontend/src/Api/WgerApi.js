import axios from 'axios';

const WGER_BASE_URL = "https://wger.de/api/v2"

    class WgerApi {
        static async request( endpoint, data = {}, method = "get"){
            console.debug("WGER Call:", endpoint, data, method)

            const url = `${WGER_BASE_URL}/${endpoint}`

            const params = (method === "get")
            ? data
            : {};
            try{
                const res = (await axios({url, method, data, params}));
                return res
            } catch(err){
                console.error("wger API Error:", err.response)
                let message = err.response.data.error.message
                throw Array.isArray(message) ? message : [message];
            }
        }

        static async getAllExercises(endpoint = 'exercisebaseinfo/'){
            if(endpoint.includes(WGER_BASE_URL)){
                endpoint = endpoint.split(`${WGER_BASE_URL}/`)[1]
            };
            
            console.log(endpoint)
            let res = await this.request(`${endpoint}`)
            console.log(res)
            return res
        }

    }


    export default WgerApi