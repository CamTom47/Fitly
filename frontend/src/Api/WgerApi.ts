import axios, { AxiosResponse } from 'axios';

const WGER_BASE_URL : string = "https://wger.de/api/v2"

    class WgerApi {
        static async request( endpoint: string, data: object = {}, method: string = "get"){
            console.debug("WGER Call:", endpoint, data, method)

            const url: string = `${WGER_BASE_URL}/${endpoint}`

            const params: {} = (method === "get")
            ? data
            : {};
            try{
                const res = (await axios({url, method, data, params}));
                return res
            } catch(err){
                console.error("wger API Error:", err.response)
                let message : string = err.response.data.error.message
                throw Array.isArray(message) ? message : [message];
            }
        }

        static async getAllExercises<Promise>(endpoint : string = 'exercisebaseinfo/'): Promise {

            // The initial request will set the next and previous calls with a end point similar to 
            // "next": "https://wger.de/api/v2/exercisebaseinfo/?limit=20&offset=20",
            // Splitting on the WGER_BASE_URL allows for the new and point to be added to the default WGER_BASE_URL

            if(endpoint.includes(WGER_BASE_URL)){
                endpoint = endpoint.split(`${WGER_BASE_URL}/`)[1]
            };
            
            let res = await this.request(endpoint)

            return res.data
        }

    }


    export default WgerApi