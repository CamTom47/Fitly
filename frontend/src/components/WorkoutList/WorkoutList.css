@media screen and (min-width: 1300px){
    #WorkoutListContainer {
        background: hsl(0, 0%, 10%);
        display: flex;
        flex-direction: column;
        align-items: left;
    }

    #WorkoutListContainer button:hover{
        background-color: hsla(212, 11%, 75%);
        box-shadow: 0px 3px 10px hsla(212, 11%, 80%);
        background-image: linear-gradient(45deg,hsla(212, 11%, 60%), hsla(212, 11%, 70%));
        border: 2px solid hsla(212, 11%, 75%);
        transition: all .35s;
    }

    .WorkoutListTitle{
        margin-top: .5rem;
        margin-left: 2rem;
        display: static;
        text-align: left;
        width: 100%;
        color: white;
    }

    .filterworkoutdivider{
        background: var(--App-background-primary);
        display: flex;
        flex-direction: row;
        justify-content: left;
        width: 100%;
    }

    .filterworkoutdividerinner{
        background: black;
        display: flex;
        flex-direction: row;
        justify-content: left;
        width: 100%;
        margin: 1rem;
        border-radius: 20px;
    }

    .filtersection{
        display: flex;
        flex-direction: column;
        align-items: center;
        background: var(--primary-gradient-light);
        width: 90%;
        border-radius: 20px 0 0 20px;   
        padding: 1rem;
        margin-left: 0;
        width: 25%;
    }

    .workouthead{
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        width: 100%;
        background-color: hsl(0, 0%, 50%);
        border-bottom: 1px solid hsl(0, 0%, 40%);
        border-radius: 20px 20px 0 0 ;
        padding: 1rem;
    }

    .workouthead2{
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        width: 100%;
        background-color: hsl(0, 0%, 50%);
        padding: 1rem;
    }

    .workoutbody{
        display: grid;
        background-color: hsl(0, 0%, 50%);
        width: 100%;
        grid-template-columns: [line1] 20rem [line2] 20rem [line3] 20rem [line4] 20rem;
        grid-template-rows: [row1] 12rem [row2] 12rem [row3] 12rem ;
        column-gap: 1rem;
    }

    .workoutsection{
       display: flex;
        flex-direction: column;
        align-items: center;
        background: var(--primary-gradient-light);
        border-radius: 0 20px 20px 0;
        border: none;
        width: 90%;
        padding: 1rem;
    }

    .searchbar input{
        border: 1px solid black;
        border-radius: 5px;
        padding: 0 .5rem 0 .5rem;
        text-align: center;
        margin-left: .5rem;
    }    

    #wgerexercisesdiv{
        display: flex;
        flex-direction: column;
        align-items: center;
    }

    .arrow{
        margin: .5rem;
    }

}
