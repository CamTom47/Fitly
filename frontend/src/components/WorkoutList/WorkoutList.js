import React, {useEffect, useState, useCallback, useContext} from "react";
import FitlyApi from "../../Api/FitlyApi"
import WorkoutSummary  from "../WorkoutSummary/WorkoutSummary"
import { v4 as uuid } from "uuid";
import { useNavigate } from "react-router-dom";
import NewWorkoutForm from "../Forms/NewWorkoutForm/NewWorkoutForm";
import UserContext from "../../context/UserContext";

//custom hooks
import useToggle from "../../hooks/useToggle/useToggle";

/**
 * WorkoutList Component => returns a user's list of workouts
 * 
 * state: workouts => array of workouts
 * 
 * props: none
 */

const WorkoutList = () => {

    const navigate = useNavigate();
    let currentUser = useContext(UserContext).currentUser;

    //Check if user is logged in. If not navigate back to the login page
    // useEffect(() => {
    //     if(currentUser === null) {
    //         navigate('/')
    //     }
    // }, [])
    
    const [workouts, setWorkouts] = useState([]);
    const [showCreateWorkoutForm, setShowCreateWorkoutForm] = useToggle();

    const callWorkouts = useCallback(async () => {
            try{
                    const workouts = await FitlyApi.findAllWorkouts();
                    setWorkouts(workouts);
            } catch(err){
                return err
            }
        } , [showCreateWorkoutForm]
    ) 

    useEffect( () => {
        callWorkouts();
    }, [callWorkouts])
    
    
    const toggleCreateForm = () => { 
        setShowCreateWorkoutForm( showCreateWorkoutForm => !showCreateWorkoutForm)
    }

    return (showCreateWorkoutForm)
    ? <NewWorkoutForm toggleCreateForm={toggleCreateForm}/>
    : ( 
        <div className="d-flex justify-content-center">
            <div className="d-flex align-items-center flex-column w-75">
                <h3>Workouts</h3>
                <hr className="w-100"></hr>
                <button className="btn btn-secondary" onClick={toggleCreateForm}>Create New Workout</button>
                <div className="d-flex flex-row justify-content-left flex-wrap column-gap-5 row-gap-5 pt-4">
                    {workouts.map(workout => (
                        <WorkoutSummary key={uuid()} workout={workout}/>)
                    )}
                </div>
            </div>
        </div>
        )
}

export default WorkoutList;