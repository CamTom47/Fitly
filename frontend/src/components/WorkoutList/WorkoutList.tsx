import React, { useEffect, useState, useCallback } from "react";
import $ from "jquery";
import WorkoutSummary from "../WorkoutSummary/WorkoutSummary";
import LoadingComponent from "../LoadingComponent/LoadingComponent";
import { v4 as uuid } from "uuid";
import useToggle from "../../hooks/useToggle/useToggle";
import NewWorkoutForm from "../Forms/NewWorkoutForm/NewWorkoutForm";
import { useAppDispatch, useAppSelector } from "../../hooks/reduxHooks";
import { findAllWorkouts, selectWorkouts } from "../../slices/workoutsSlice";
import { findAllCategories, selectCategories } from "../../slices/categoriesSlice";
import { findAllCircuits } from "../../slices/circuitsSlice";

import "./WorkoutList.css";

import { workoutFilter } from "../../helpers/filters";
import { findAllExercises } from "../../slices/exercisesSlice";

interface WorkoutQuery {
	category?: String;
	favorited?: Boolean;
}
interface WorkoutSortBy {
	name?: String;
	category?: String;
	timesCompleted?: Number;
	lastCompleted?: Date;
}

const WorkoutList = (): React.JSX.Element => {
	const initialQueryState = {};
	const initialSortByState = {};

	const dispatch = useAppDispatch();
	const workoutList = useAppSelector(selectWorkouts);
	const categories = useAppSelector(selectCategories);
	
	const [showCreateWorkoutForm, setShowCreateWorkoutForm] = useToggle();
	const [isLoading, setIsLoading] = useState(true);
	const [filterTerms, setFilterTerms] = useState<WorkoutQuery>(initialQueryState);
	const [searchTerm, setSearchTerm] = useState<String | null>(null);
	const [sortBy, setSortBy] = useState<{}>(initialSortByState);

	dispatch(findAllExercises);

	let workouts = workoutList.filter((workout) => {
		if (searchTerm) return workout.name.toLowerCase().includes(searchTerm.toLowerCase());
		else return workoutList;
	});

	useEffect(() => {
		dispatch(findAllCategories());
		dispatch(findAllCircuits());
	}, []);

	const getWorkoutInfo = useCallback( async () => {
		dispatch(findAllWorkouts({ filterBy: filterTerms, sortBy }));
		setIsLoading(false);
	}, [showCreateWorkoutForm, filterTerms, sortBy]);

	const handleFilter = (e) => {
		$(`input[name=${e.target.name}`).not(`input[value="${e.target.value}"]`).prop("checked", false);
		setFilterTerms(workoutFilter(filterTerms, e));
	};

	useEffect(() => {
		getWorkoutInfo();
	}, [getWorkoutInfo]);

	const toggleCreateForm = () => {
		setShowCreateWorkoutForm();
	};

	const workoutSummaryComponents = workouts.map((workout, idx) => (
		<WorkoutSummary key={uuid()} workoutNumber={idx} workout={workout} />
	));

	const categoryFilterComponents = categories.map((category) => (
		<div className='WorkoutFilterDiv'>
			<label htmlFor='categoryFilter'>{category.name}</label>
			<input
				id={category.id}
				className='filterButton'
				onClick={handleFilter}
				type='checkbox'
				name='categoryFilter'
				value={category.name}
			/>
		</div>
	));

	const favoriteFilterComponent = (
		<div className='WorkoutFilterDiv'>
			<label htmlFor='favoriteFilter'>Favorited</label>
			<input className='filterButton' onClick={handleFilter} type='checkbox' name='favoritedFilter' value='favorited' />
		</div>
	);

	const resetFilter = () => {
		setFilterTerms(initialQueryState);
		$(".filterButton").prop("checked", false);
	};

	const handleSearch = (e) => {
		setSearchTerm(e.target.value);
	};

	const handleSort = (e) => {
		let sortType = e.target.selectedOptions[0].dataset["type"];
		let sortObj = {}
		sortObj[sortType] = e.target.value
		setSortBy(sortObj);
	};

	return isLoading ? (
		<LoadingComponent />
	) : showCreateWorkoutForm ? (
		<NewWorkoutForm toggleCreateForm={toggleCreateForm} />
	) : (
		<div id='WorkoutListContainer'>
			<div className='filterWorkoutDivider'>
				<div className='filterWorkoutDividerInner'>
					<div className='filtersection'>
					{Object.keys(filterTerms).length !== 0 && <button onClick={resetFilter}>Clear Filter</button>}

						<p>Category</p>
						<form className='FilterForm'>{categoryFilterComponents}</form>
						<p>Favorites</p>
						<form className='FilterForm'>{favoriteFilterComponent}</form>
					</div>
					<div className='workoutsection'>
						<div className='workouthead'>
							<div className='searchbar'>
								<form>
									<input type='text' placeholder='Find Workout' onChange={handleSearch}></input>
								</form>
							</div>
							<div>
								<form>
									<select className="sortForm" onChange={handleSort} name='sort'>
										<option value=''>Sort By</option>
										<option data-type="name" value='nameAsc'>Workout Name A-Z</option>
										<option data-type="name" value='nameDesc'>Workout Name Z-A</option>
										<option data-type="category" value='categoryAsc'>Type Of Workout A-Z</option>
										<option data-type="category" value='categoryDesc'>Type Of Workout Z-A</option>
										<option data-type="timesCompleted" value='timeCompletedAsc'>Times Completed 0-9</option>
										<option data-type="timesCompleted" value='timeCompleteDesc'>Times Completed 9-0</option>
										<option data-type="lastCompleted" value='lastCompletedAsc'>Last Completed 0-9</option>
										<option data-type="lastCompleted" value='lastCompletedDesc'>Last Completed 9-0</option>
									</select>
								</form>
							</div>
							<button id='addWorkoutButton' onClick={toggleCreateForm}>
								+
							</button>
						</div>
						<table className='workoutTable'>
							<tr>
								<th></th>
								<th>#</th>
								<th>Workout Name</th>
								<th>Type Of Workout</th>
								<th>Times Completed</th>
								<th>Last Completed</th>
							</tr>
							{workoutSummaryComponents}
						</table>
					</div>
				</div>
			</div>
		</div>
	);
};

export default WorkoutList;
