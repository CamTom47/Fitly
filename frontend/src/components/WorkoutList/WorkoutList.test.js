import { render, waitFor} from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
// import userEvent from '@testing-library/user-event'
import FitlyApi from '../../Api/FitlyApi';
import WgerApi from '../../Api/WgerApi';
import UserProvider from '../../../testUtils';
import WorkoutList from './WorkoutList';
 
//Mock FitlyAPI
jest.mock('../../Api/FitlyApi', () => ({
    findAllWorkouts: jest.fn(() => (
        [{
         id: 1,
        user_id: 1,
        name: "test workout",
        category: 1,
        favorited: false
    },{
        id: 2,
       user_id: 2,
       name: "test workout 2",
       category: 2,
       favorited: false
   }]
    ))
}));

//Mock WgerApi
jest.mock('../../Api/WgerApi', () => {
});



describe('WorkoutList Functionality', () => {

    let workouts;

    waitFor(async () => {
        workouts = await FitlyApi.findAllWorkouts();
    })
    let children = (
        <WorkoutList workouts={workouts}/>
    );

    


  it('it renders without crashing', async () => {

    render(
        <UserProvider children={children}/>
      );
    }); 
    
    
    it('matches snapshot', async () => {
        const {asFragment} = render(
    <UserProvider children={children}/>
)
     expect(asFragment()).toMatchSnapshot();
    })
});