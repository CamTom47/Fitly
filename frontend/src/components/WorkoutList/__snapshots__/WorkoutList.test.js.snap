// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutList Functionality matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="d-flex align-items-center flex-column w-75"
    >
      <h3>
        Workouts
      </h3>
      <hr
        class="w-100"
      />
      <button
        class="btn btn-secondary"
      >
        Create New Workout
      </button>
      <div
        class="d-flex flex-row justify-content-left flex-wrap column-gap-5 row-gap-5 pt-4"
      />
    </div>
  </div>
</DocumentFragment>
`;
